packages/aws-cdk/test/api/hotswap/s3-bucket-hotswap-deployments.test.ts

```diff
@@ -0,0 +1,692 @@
+        S3Deployment: {
+          Type: 'Custom::CDKBucketDeployment',
+          Properties: {
+            ServiceToken: 'a-lambda-arn',
+            SourceBucketNames: ['src-bucket'],
+            SourceObjectKeys: ['src-key-new'],
+            DestinationBucketName: 'dest-bucket',
+            DestinationBucketKeyPrefix: 'my-key/some-new-prefix',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await hotswapMockSdkProvider.tryHotswapDeployment(cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).not.toBeUndefined();
+
+  expect(mockLambdaInvoke).toHaveBeenCalledWith({
+    FunctionName: 'a-lambda-arn',
+    Payload: JSON.stringify({
+      ...payloadWithoutCustomResProps,
+      ResourceProperties: {
+        SourceBucketNames: ['src-bucket'],
+        SourceObjectKeys: ['src-key-new'],
+        DestinationBucketName: 'dest-bucket',
+        DestinationBucketKeyPrefix: 'my-key/some-new-prefix',
+      },
+    }),
+  });
+});
+
+test('does not call the invoke() API when a resource with type that is not Custom::CDKBucketDeployment but has the same properties is changed', async () => {
+  // GIVEN
+  setup.setCurrentCfnStackTemplate({
+    Resources: {
+      S3Deployment: {
+        Type: 'Custom::NotCDKBucketDeployment',
+        Properties: {
+          ServiceToken: 'a-lambda-arn',
+          SourceBucketNames: ['src-bucket'],
+          SourceObjectKeys: ['src-key-old'],
+          DestinationBucketName: 'dest-bucket',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = setup.cdkStackArtifactOf({
+    template: {
+      Resources: {
+        S3Deployment: {
+          Type: 'Custom::NotCDKBucketDeployment',
+          Properties: {
+            ServiceToken: 'a-lambda-arn',
+            SourceBucketNames: ['src-bucket'],
+            SourceObjectKeys: ['src-key-new'],
+            DestinationBucketName: 'dest-bucket',
+          },
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await hotswapMockSdkProvider.tryHotswapDeployment(cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).toBeUndefined();
+  expect(mockLambdaInvoke).not.toHaveBeenCalled();
+});
+
+test('does not call the invokeLambda() api if the updated Policy has no Roles', async () => {
+  // GIVEN
+  setup.setCurrentCfnStackTemplate({
+    Resources: {
+      S3Deployment: {
+        Type: 'Custom::CDKBucketDeployment',
+        Properties: {
+          ServiceToken: 'a-lambda-arn',
+          SourceBucketNames: ['src-bucket'],
+          SourceObjectKeys: ['src-key-old'],
+          DestinationBucketName: 'dest-bucket',
+        },
+      },
+      Policy: {
+        Type: 'AWS::IAM::Policy',
+        Properties: {
+          PolicyDocument: {
+            Statement: [
+              {
+                Action: ['s3:GetObject*'],
+                Effect: 'Allow',
+                Resource: {
+                  'Fn::GetAtt': [
+                    'WebsiteBucketOld',
```

Comment:
Hmm. It's kind of weird that this is an invalid template. Any reason we can't test this using an Asset parameter, for example?

Like we do [here](https://github.com/aws/aws-cdk/blob/8e3269b11ae013c74fd46046697ebd4d647bb7ed/packages/aws-cdk/test/api/hotswap/lambda-hotswap-deployments.test.ts#L153-L211), for instance.